src/queries/getPackageData.ts

```diff
@@ -13,14 +13,26 @@ function isCompatible(nodeVersion: string, depRange: string) {
   // if a dependency has `*` for the node version, it's always compatible
   if (['x', '*'].includes(depRange)) return true;
 
-  let compatible;
+  return depRange
+    .split('||')
+    .map((range) => removeWhitespace(range))
+    .some((range) => safeSatisfies(nodeVersion, range));
+}
+
+// accounts for `AND` ranges -- ie, `'>=1.2.9 <2.0.0'`
+function safeSatisfies(nodeVersion: string, range: string) {
+  return range.split(' ').every((r) => satisfies(nodeVersion, r));
+}
 
-  const logicalOrRegEx = /\|\|/;
-  if (depRange && logicalOrRegEx.test(depRange)) {
-    const rangeArray = depRange.split('||').map((range) => range.replaceAll(' ', ''));
-    compatible = rangeArray.some((range) => satisfies(nodeVersion, range));
-  } else {
-    compatible = satisfies(nodeVersion, depRange.replaceAll(' ', ''));
-  }
-  return compatible;
+// trims leading and trailing whitespace, whitespace
+// between the comparator operators and the actual version number, 
+// and whitespace between the numbers/wildcards/decimals in the actual
+// version number. ie, ' > = 1 2. 0 .0 ' becomes '>=12.0.0'
+//
+// it also ensures there's only one space in logical `AND` ranges
+// ie, '>=1.2.9 <2.0.0', because we want to split those later
+function removeWhitespace(range: string) {
+  const extraSpaceRegEx =
+    /((?<=(<|>))(\s+)(?=(=)))|(?<=(<|>|=|\^|~))(\s+)(?=\d)|((?<=(\d|\.|\*|x|X))(\s+)(?=(\d|\.|\*|x|X)))|(?<=\d)(\s+)(?=\s<|>)/g;
```

Comment:
this looks scary xD, I'd like to give it a try into changing this to something less scary, can you share some of the malformed strings and the expected output?

it might be a good idea to add tests for this function with input/output examples

cause I can imagine there's some other solution (maybe combining regexp with strip, or smaller regexps to replace different parts one at a time)

Reply:
hi lew :D

Reply:
unfortunately, JavaScript doesn't have a native `strip` for strings. we could break down the RegEx into each conditional one though and then just chain `replace` calls?

```typescript
const comparatorWhitespace = /((?<=(<|>))(\s+)(?=(=)))/g;
const comparatorAndVersionWhiteSpace = /(?<=(<|>|=|\^|~))(\s+)(?=\d)/g;
const versionWhitespace = /((?<=(\d|\.|\*|x|X))(\s+)(?=(\d|\.|\*|x|X)))/g;
const betweenWhitespace = /(?<=\d)(\s+)(?=\s<|>)/g;

return range
  .trim()
  .replace(comparatorWhitespace, '')
  .replace(comparatorAndVersionWhiteSpace, '')
  .replace(versionWhitespace, '')
  .replace(betweenWhitespace, '');
```

also, here are some test cases for things that would cause `satisfies` to throw:

```
> =12.0.0
> 12.0.0
1 2. 0 .0
>= 12.0.0
> = 1 2 . 0 . 0
> = 4 . 2 . 0     < 8. 0 .0
```

Reply:
oohhh, i just came up with a waaay shorter RegEx using _negative_ lookahead:

```
/(\s)(?!(<|>|^|~))/g
```
That matches any whitespace that is not immediately followed by `< | > | ~ | ^`

I don't think it's perfect (because I'm not sure if you can do an AND range like `>=4.2.0 8.0.0`, in which case this would remove the whitespace between them), but I should cover all the cases in the above comment.

Reply:
Hey @kindoflew, we definitely need to move forward with a fully working solution, as this `/(\s)(?!(<|>|^|~))/g` is not. Let's consider the original one you suggested. Also, did you have a chance to measure the execution speed between the call replace method with one long regex and a couple of calls with the separated? I think it is a vital part of the solution to be fast because we may deal with tons of dependencies.

Reply:
@KostiantynPopovych, i'm _pretty_ sure that both RegExes will do the same thing -- neither one would cover something like `>=4.2.0 8.0.0`.

basically, we'd have to write another RegEx for SemVer ranges (which [isn't exactly straightforward](). and those are just for version numbers -- they don't include comparators) and then strip all but one whitespace between them.

i think i'm sort of leaning towards the shorter one, since it'll (hopefully) be a temporary fix until `compare-versions` releases the next major version (or `npm`'s `semver` package figures out how to get rid of it's cyclical dependency and we could just use that one instead -- it has all this stuff out of the box).

Reply:
@kindoflew Oh, I see, I was playing with these solutions yesterday and probably messed something up because the original (long) regex was working for this `>=4.2.0 8.0.0` case, maybe I have fixed one thing but crashed another. So anyway, if we decide to move with any of them, we need to inform users about this issue, probably creating a section in `README.md` named `Known Issues` and describing this behavior there.

How does it sound to you?

Reply:
@KostiantynPopovych i figured out one last thing to make `'>=4.2.0     8.0.0'` work.

in `safeSatisfies` after we `split(' ')`, we filter out any extra whitespace between them (`filter((r) => !!r)`) and that should leave us with two individual ranges, no matter what. the whole thing looks like:

```javascript
range
  .split(' ')
  .filter((r) => !!r)
  .every((r) => satisfies(nodeVersion, r))
```

i also did a very small amount of testing execution speed between long regex, split long regex, and the short regex. they all seemed to pretty consistently take between 12ms and 17ms (with long regex staying closer to 12ms almost every time). all of this was tested on the dependencies of `depngn` (using `console.time`). short regex and split long had some times between 15ms and 18ms _slightly_ more often than long regex, but these weren't very rigorous tests -- i think statistically they are all about the same.

i'm down to do whatever you'd like! i'm leaning towards long regex, with maybe a comment about what each chunk represents? either way, that `filter` i mentioned solves the last whitespace problem, so 🎉

Reply:
@kindoflew, sorry for the long reply! In the case of adding an additional filter - `range.trim()` sounds redundant, coz we will filter these white spaces in the following function, isn't it?

Reply:
oh yeah, you're right. i thought i solved something else but i didn't 😅. i have too many versions of this locally.

anyway, the issue isn't with `>=4.2.0 8.0.0`, it's with `>=4.2.0      8.0.0`.

but, i think i finally figured it out (i didn't benchmark execution time, but with the optimizations in #39, i don't think it'll matter much. plus i've been looking at how big libraries like `axios` and `semver` do regex stuff, and they seem to have multiple `replace` and `match` calls all over the place and do just fine).

anyways, this (hopefully) is the final version i came up with:

```typescript
function removeWhitespace(range: string) {
  const comparatorWhitespace = /((?<=(<|>))(\s+)(?=(=)))/g;
  const comparatorAndVersionWhiteSpace = /(?<=(<|>|=|\^|~))(\s+)(?=\d)/g;
  const digitAndDecimalWhiteSpace =
    /((?<=(\d))(\s+)(?=(\.|\*|x|X)))|((?<=(\.|\*|x|X))(\s+)(?=(\d)))/g;
  const remainingRange =
   /(((\d\s*){1,3})\.\s*(\d\s*){1,3}\.\s*((0\s*)|(\d\s*){1,2})(?!(((<|>)=?)|~|^)?(((\d\s*){1,3})\.)|(\d)))/g;
  return range
    .trim()
    .replace(comparatorWhitespace, '')
    .replace(comparatorAndVersionWhiteSpace, '')
    .replace(remainingWhitespace, (match: string) => match.replace(/\s+/g, ''));
}
```

The first two regexes (`comparatorWhitespace` and `comparatorAndVersionWhitespace`) strip out the whitespace between what the names say. this would leave us with things like:

```
>=4.2.0      >=18.0.0
>=4.2. 0     1 8 .0.0   
>=1 2.  0.0
>1 2.0.1 0    1
1 2 . 1 0 . 0 
>=1 2.0.1 0
>=1   2.0.0
>=4.2.0      8.0
>=4.2.1  0      8
```

The last regex (`remainingRange`) matches a single SemVer version that may or may not have whitespace between digits or digits and decimals. and then we replace it with the same string with whitespace removed. this leaves us with either a whitespace-less single range or an AND range with a single space in between. we still need the last `filter` call because the last regex will not match on something like `>1    <3`, which is a valid range (minus the extra whitespace).

Reply:
@KostiantynPopovych however, looking at what `node-semver` does, they assume the only additional spaces will be between comparator operator characters or between comparator operators and the version number -- they consider additional space between digits/decimals to be malformed and just throw an error.

we could do that (but instead of throwing, maybe we just return `{ compatible: undefined, range: {malformedRange} }` or maybe even `{ compatible: 'malformed', range: {malformedRange} }`. what do you think of that?

Reply:
i think because the RegEx to parse it is impossible? or at least computationally expensive.

consider the case where you have the following, malformed range:

```
<1   2
```

is that supposed to be "less than 12" or "less than 1 AND exactly 2"? i think because there are so many different variations for SemVer strings, you kind of have to follow the happy path for the most part and just throw when you can't be sure. that's my assumption from researching this stuff the past few weeks ¯\\\_(ツ)_/¯

Reply:
Sounds reasonable) So, I believe we can go with this approach.